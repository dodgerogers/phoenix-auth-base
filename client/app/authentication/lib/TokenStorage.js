import Cookies from 'js-cookie';
import moment from 'moment';


export const COOKIE_KEY = 'token';
export const generateExpiry = token => moment(token.createdAt).add(token.expiresIn, 'seconds').toDate();
export const encode = token => btoa(JSON.stringify(token));
export const decode = token => {
  try {
    return JSON.parse(atob(token));
  } catch (err) {
    return null;
  }
};

export const store = token => {
  return new Promise((resolve, reject) => {
    const expires = generateExpiry(token);
    const cookie = Cookies.set(COOKIE_KEY, encode(token), { expires });

    if (cookie) {
      return resolve(cookie);
    }

    reject(null);
  });
}

export const fetch = () => {
  return new Promise((resolve, reject) => {
    const encodedToken = Cookies.get(COOKIE_KEY);

    if (encodedToken) {
      const decodedToken = decode(encodedToken);
      return resolve(decodedToken);
    }

    reject(null);
  });
};

export const remove = () => {
  return new Promise((resolve, reject) => {
    try {
      const result = Cookies.remove(COOKIE_KEY);
      resolve(result);
    } catch (err) {
      reject(err);
    }
  });
};
