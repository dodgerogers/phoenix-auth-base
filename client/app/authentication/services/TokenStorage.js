import moment from 'moment';


export const STORAGE_KEY = 'token';
const storage = global.localStorage;

export const encode = token => btoa(JSON.stringify(token));
export const generateExpiry = token => moment(token.createdAt).add(token.expiresIn, 'seconds').toDate();

export const decode = token => {
  try {
    return JSON.parse(atob(token));
  } catch (err) {
    return null;
  }
};

export const store = token => {
  return new Promise((resolve, reject) => {
    try {
      storage.setItem(STORAGE_KEY, encode(token));
      resolve();
    } catch (err) {
      reject(err);
    }
  });
}

export const fetch = () => {
  return new Promise((resolve, reject) => {
    const foundToken = storage.getItem(STORAGE_KEY);
    if (foundToken === null) {
      return reject(null);
    }

    const decodedToken = decode(foundToken);
    return resolve(decodedToken);
  });
};

export const remove = () => {
  return new Promise((resolve, reject) => {
    try {
      storage.removeItem(STORAGE_KEY);
      resolve(null);
    } catch (err) {
      reject(err);
    }
  });
};
