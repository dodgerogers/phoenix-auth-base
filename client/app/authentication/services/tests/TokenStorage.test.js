import Cookies from 'js-cookie';
import * as TokenStorage from '../TokenStorage';

describe('Token', () => {
  const token = () => ({
    accessToken: 'a1b2c3d4e5f6',
    expiresIn: 900,
    createdAt: '2000-12-12T00:01:01',
  });

  const encodedToken = 'eyJhY2Nlc3NUb2tlbiI6ImExYjJjM2Q0ZTVmNiIsImV4cGlyZXNJbiI6OTAwLCJjcmVhdGVkQXQiOiIyMDAwLTEyLTEyVDAwOjAxOjAxIn0=';

  describe('COOKIE_KEY', () => {
    it('matches snapshot', () => {
      expect(TokenStorage.COOKIE_KEY).toMatchSnapshot();
    });
  });

  describe('encode', () => {
    it('encodes token object into expected base64 string', () => {
      expect(TokenStorage.encode(token())).toEqual(encodedToken);
    });
  });

  describe('decode', () => {
    it('decodes previously encoded token successfully', () => {
      expect(TokenStorage.decode(encodedToken)).toEqual(token());
    });

    it('returns null when encoded token is not valid base64', () => {
      expect(TokenStorage.decode('invalid')).toEqual(null);
    });
  });

  describe('generateExpiry', () => {
    it('constructs date from the created_at and expires_in properties', () => {
      expect(TokenStorage.generateExpiry(token())).toMatchSnapshot();
    });
  });

  describe('store', () => {
    it('should resolve truthy when able to set cookie', () => {
      const cookie = 'token cookie';
      Cookies.set = jest.fn(() => cookie);

      return TokenStorage.store(token())
        .then(result => {
          expect(Cookies.set.mock.calls).toMatchSnapshot();
          expect(result).toEqual(cookie);
        });
    });

    it('should resolve falsey when not able to set cookie', () => {
      Cookies.set = jest.fn(() => null);

      return TokenStorage.store(token())
        .catch(result => {
          expect(Cookies.set.mock.calls).toMatchSnapshot();
          expect(result).toEqual(null);
        });
    });
  });

  describe('fetch', () => {
    it('should resolve with the decoded token when Cookies.get is successful', () => {
      Cookies.get = jest.fn(() => encodedToken);

      return TokenStorage.fetch()
        .then(result => {
          expect(result).toEqual(token());
        });
    });

    it('should reject with false when Cookies.get fails', () => {
      Cookies.get = jest.fn(() => null);

      return TokenStorage.fetch()
        .catch(result => {
          expect(result).toEqual(null);
        });
    });
  });

  describe('remove', () => {
    it('should resolve promise with null when Cookies.remove is successful', () => {
      Cookies.remove = jest.fn(() => null);

      return TokenStorage.remove()
        .then(result => {
          expect(Cookies.remove.mock.calls).toMatchSnapshot();
          expect(result).toEqual(null);
        });
    });

    it('should reject promise with null when Cookies.remove fails', () => {
      const error = 'Something went wrong';
      Cookies.remove = jest.fn(() => {
        throw(error);
      });

      return TokenStorage.remove()
        .catch(result => {
          expect(Cookies.remove.mock.calls).toMatchSnapshot();
          expect(result).toEqual(error);
        });
    });
  });
});
