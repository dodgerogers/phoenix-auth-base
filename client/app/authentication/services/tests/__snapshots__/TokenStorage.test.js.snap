// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenStorage COOKIE_KEY matches snapshot 1`] = `undefined`;

exports[`TokenStorage generateExpiry constructs date from the created_at and expires_in properties 1`] = `2000-12-12T05:16:01.000Z`;

exports[`TokenStorage remove should reject promise with null when storage.remove fails 1`] = `
Array [
  Array [
    "token",
  ],
]
`;

exports[`TokenStorage remove should resolve promise with null when storage.remove is successful 1`] = `
Array [
  Array [
    "token",
  ],
]
`;

exports[`TokenStorage store should resolve truthy when able to set token 1`] = `
Array [
  Array [
    "token",
    "eyJhY2Nlc3NUb2tlbiI6ImExYjJjM2Q0ZTVmNiIsImV4cGlyZXNJbiI6OTAwLCJjcmVhdGVkQXQiOiIyMDAwLTEyLTEyVDAwOjAxOjAxIn0=",
  ],
]
`;
